### ISSUE TITLE
Support running Dask with Istio / Envoy proxy

### ISSUE THREAD

**Issue by elyast** [dask/dask-kubernetes#197]
In order to run Dask with Envoy enabled, each worker needs to have service created and expose it through `contact-address`, I was wondering if we can add creation of a service to the worker through a worker-template, e.g. not assume its a single pod definition inside a template, but consider service definition as well if defined ?

**Comment by jacobtomlinson** [comment]
I would love to know more about your use case and what benefits you are hoping to get by using Envoy with Dask.

Currently workers communicate with each other and the scheduler using the standard container network. They are able to discover their own IP and then share that with the scheduler during registration. 

If we decided to put workers behind services within the cluster they would need to be able to discover that address instead and pass that along to the scheduler.

Dask communication is also done with a custom RPC protocol over TCP and so these services would have to support L4 routing. I don't know enough about Envoy to advise on this yet.

In the latest release, we added support for the scheduler running as a separate pod and creating a service for it to enable external access. So something similar could be done for workers if that is desirable.

**Comment by jacobtomlinson** [comment]
**TL;DR** from #200 

> Istio blocks communication over IP, Dask wants to communicate over IP, it fails.

> With latest Istio it looks like IP in `serviceentries` won't work.

> So looks like we/I need to find a way to start using services [for worker pods].

**Comment by LeonardAukea** [comment]
This would be great if one would want to use dask kubernetes with kubeflow (Uses istio all over the place).

**Comment by sshakeri** [comment]
Meanwhile if you have to, a workaround for Kubeflow is to disable auto side injection and bring the jupyter pod back up.  After you kill the jupyter pod it will come back without the sidecar you can create`KubeCluster` 
```
kubectl label namespace projects istio-injection-
kubectl delete pod your_jupyter
```

**Comment by Kyrremann** [comment]
Just remember that this also removed the security/login that Istio/Dex gives you.

**Comment by LeonardAukea** [comment]
@sshakeri  I don't have to. I just think it would be great To start working towards getting dask running nicely with kubeflow.

**Comment by alexifm** [comment]
Wrote this on #200 but since this issue is specifically about istio:

I did a bit of digging and it seem that more recent versions of istio have made substantial changes that could allow Dask to work:
https://medium.com/@dinup24/istio-setting-up-ip-address-based-access-control-d16bac59b2d3

Has anyone else tried Dask with istio recently?  (We're about to)

**Comment by dirtyValera** [comment]
@alexifm 

Curious what your result with Dask + Istio are? I'm trying to set up Kubeflow with Dask as well, any info is appreciated. Thanks!

**Comment by haf** [comment]
I'd also be interested in this; Istio is a mesh that enables us to configure cross-company security policies and monitor and shape traffic. It would be useful to have the monitoring abilities also for Dask.

**Comment by tikr7** [comment]
I used the [KubeCluster](https://kubernetes.dask.org/en/latest/kubecluster.html) way so far with Kubeflow (of course including Istio) in a JupyterLab notebook.

I got it working by turning off Istio on the Dask cluster (workarounds https://github.com/dask/dask-kubernetes/issues/361)
```
vi worker-spec.yml
kind: Pod
metadata:
  annotations:
    sidecar.istio.io/inject: false
...
```
and by opening port `8786` in the JupyterLab, it solves the problem `RuntimeError: Cluster failed to start. Timed out during handshake while connecting to tcp://<JupyterLab-notebook-name>-jovyan-f9dd5594-0.<namespace>:8786 after 30 s`
```
kubectl edit statefulset -n <namespace> <JupyterLab-notebook-name>
spec:
  template:
    metadata:
      annotations:
        traffic.sidecar.istio.io/excludeOutboundPorts: "8786"
```
But I would call it a workaround because when turning off Istio, all the nice feature like monitoring, tracebility, security, ... are gone as well.


Are there any other ideas?
Are there any news?
Anything planned to improve Dask@Istio?

**Comment by jacobtomlinson** [comment]
We are addressing Kubeflow (and Istio) support as part of #392

**Comment by tikr7** [comment]
Could you give us a short status? Would be very nice!

**Comment by jacobtomlinson** [comment]
@tikr7 the operator is a work in progress. We are working towards a deadline of end of May, but expect to have things to show way before then. If you want any more information please see the PR.

**Comment by jacobtomlinson** [comment]
I'm to close this issue out now. We have a workaround for the classic `KubeCluster` with https://github.com/dask/dask-kubernetes/issues/197#issuecomment-1048593317 and do not intend to ever add true support there.

The operator is the path forwards and are tracking Istio support there in #482 where we have an implementation plan.
